plugins/kubernetes/app/models/watchers/base_watcher.rb

```diff
@@ -8,19 +8,20 @@ class BaseWatcher
     def initialize(watch_stream, log: true)
       @watch_stream, @log, = watch_stream, log
       @start_time = Time.now
+      @enabled = false
     end
 
     def start_watching(&block)
+      @enabled = true
       @watch_stream.each do |notice|
         handle_notice(notice, &block)
+        break if stop_watching?
       end
+      cleanup
     end
 
     def stop_watching
-      if @watch_stream
-        @watch_stream.finish
-        @watch_stream = nil
-      end
+      @enabled = false
     end
```

Comment:
@jonmoter in that case from the other thread you should call finish to force the connection to drop (IIRC you can safely call finish multiple times if you protect from races)